core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java

```diff
@@ -2927,4 +3018,96 @@ private static BsonDocument withNewVersion(
+    CANNOT_PARSE_REMOTE_VERSION_MESSAGE("got a remote document that could not have its version "
+        + "info parsed"),
+    DELETE_FROM_REMOTE_MESSAGE("deleting local as there are no local pending writes"),
+    DUPLICATE_KEY_EXCEPTION_MESSAGE("duplicate key exception on insert"),
+    EMPTY_VERSION_MESSAGE("remote or local have an empty version"),
+    EMPTY_UPDATE_DESCRIPTION("local change event update description is empty for UPDATE"),
+    EXCEPTION_DELETE("exception on delete: %s"),
+    EXCEPTION_INSERT("exception on insert: %s"),
+    EXCEPTION_REPLACE("exception on replace: %s"),
+    EXCEPTION_UPDATE("exception on update: %s"),
+    EXPECTED_LOCAL_DOCUMENT_TO_EXIST_MESSAGE("expected document to exist for local update change "
+        + "event"),
+    PENDING_WRITE_DELETE_MESSAGE("remote delete but a write is pending"),
+    PENDING_WRITE_EMPTY_VERSION_MESSAGE("remote or local have an empty version but a write is "
+        + "pending"),
+    PENDING_WRITE_INSTANCE_ID_MISMATCH_MESSAGE("pending write and remote events created by "
+        + "different devices"),
+    PROBABLY_GENERATED_BY_US_MESSAGE("remote change event was generated by us"),
+    REMOTE_UPDATE_WITHOUT_VERSION_MESSAGE("remote document changed but version was unmodified"),
+    SIMULTANEOUS_WRITES_MESSAGE("has multiple events at same logical time"),
+    STALE_LOCAL_WRITE_MESSAGE("remote event version has higher counter than local pending write"),
+    STALE_EVENT_MESSAGE("remote change event is stale"),
+    UNKNOWN_OPTYPE_MESSAGE("unknown operation type: %s"),
+    UNKNOWN_REMOTE_PROTOCOL_VERSION_MESSAGE("got a remote document with an unsupported "
+        + "synchronization protocol version %d"),
+    VERSION_DIFFERENT_DELETED_DOC_MESSAGE("version different on removed document"),
+    VERSION_DIFFERENT_REPLACED_DOC_MESSAGE("version different on replaced document or document was "
+        + "deleted"),
+    VERSION_DIFFERENT_UNPROCESSED_EVENT_MESSAGE("version different on unprocessed change event for "
+        + "document"),
+    VERSION_DIFFERENT_UPDATED_DOC_MESSAGE("version different on updated document or document was "
+        + "deleted");
+
+
+    static final String BASE_MESSAGE = "t='%d': %s ns=%s documentId=%s ";
+    static final String R2L_METHOD = "syncRemoteChangeEventToLocal";
+    static final String L2R_METHOD = "syncLocalToRemote";
+
+    final String message;
+
+    SyncMessage(final String message) {
+      this.message = BASE_MESSAGE.concat(message);
+    }
+
+    static String constructMessage(
+        final SyncAction action,
+        final SyncMessage message,
+        final Exception ex) {
+      final String messageWithDescription = action.applyDescription(message);
+
+      if (ex != null) {
+        return String.format(messageWithDescription, ex);
+      }
+
+      return messageWithDescription;
+    }
+
+    @Override
+    public String toString() {
+      return this.message;
+    }
+  }
+
+  private enum SyncAction {
```

Comment:
[nit] We should consider moving this to its own file.

Reply:
I'd be more inclined to say that we need to break this whole class up into about 6 different files, but I think that can wait for now considering I'm doing this under time crunch for a particular use case